### PR TITLE
Refactor/service data task

### PR DESCRIPTION
**Pull request by kpcwatson**
Add ability to use `dataTask` instead of `downloadTask` when using `FetchServiceResponseOperation`.  Initialize a `ServiceRequest` with a `ServiceRequestMethod` of either `.data` or `.download`; the default `ServiceRequestMethod` is `.download`.

Using `dataTask` means HTTP caching and ETag/Last-Modified should be handled by `URLSession` automatically.

The main impact to existing clients is that `cacheFile` is removed from  `ProcessServiceResponseOperation`'s `init`. The `init` now takes an optional `responseData`. 

I made the operations `public` instead of `open` because it seems like none of the logic was overridable anyway (except `ProcessServiceResponseOperation`). I can revert those changes, if needed.

### CODE REVIEW

File: Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift [hunk 1]
```diff
@@ -4,122 +4,185 @@
...
+    public private(set) var responseData: Data?
+    
+    // MARK: - Initialization
+    
+    public init(request: ServiceRequest, session: ServiceSession? = nil, urlSession: URLSession = URLSession.shared) {
+        self.request = request
+        
+        super.init(operations: [])
+        
+        guard let urlRequest = request.makeURLRequest(session),
```
Comment (kpcwatson) [inline Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift]: make this throw when #10 is merged

File: Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift [hunk 2]
```diff
@@ -4,122 +4,185 @@
...
-                    guard let weakSelf = self else { return }
-
-                    weakSelf.downloadFinished(url, response: response as? HTTPURLResponse, error: error as NSError?)
-                })
-
-                let taskOperation = URLSessionTaskOperation(task: task)
-                addOperation(taskOperation)
-            }
+        
+        guard let urlRequest = request.makeURLRequest(session),
```
Comment (kpcwatson) [inline Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift]: make this throw when #10 is merged

File: Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift [hunk 3]
```diff
@@ -128,116 +191,83 @@ open class DownloadServiceResponseOperation: GroupOperation {
...
  example, in a pipeline of processors).
-*/
+ */
 public enum ServiceResponseProcessorParam {
     /// Initial input is typically a memory-efficient `NSInputStream`.
     case stream(InputStream)
-
+    
     /// `NSData` can be used to pipe data from one processor to the next.
-    case data(String, Data)
```
Comment (kpcwatson) [inline Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift]: I'm not sure what the `String` parameter was used for... it wasn't used anywhere. I can add it back if it serves a purpose.

File: Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift [hunk 4]
```diff
@@ -49,73 +71,114 @@ open class FetchServiceResponseOperation: GroupOperation {
...
+        
+        guard let urlRequest = request.makeURLRequest(session),
+            let url = urlRequest.url else {
+                return
+        }
+        
+        name = "Download Service Request Operation \(url)"
+        
+        let task = urlSession.downloadTask(with: urlRequest) { [weak self] (url, _, error) in
+            self?.downloadFinished(url, error: error)
```
Comment (ktatroe) [inline Horatio/Horatio/Classes/Services/ServiceRequestOperation.swift]: Eventually, I'd like to make this take a DownloadResult (that is an enum of `fetchContent(URL)`, `failed(error)`), but I'm working on that revamp in a PR that will be coming soon, for throughout the framework).

**Comment by kpcwatson** [comment]
I'm wondering if the default parameter of `.download` should be removed so clients have to explicitly decide which they want to use. I made it default to `.download` so nothing changes for existing clients. However, my opinion is that the default should be `.data`... so maybe we should just remove the default parameter altogether.

**Comment by ktatroe** [comment]
I was wondering the same myself. I think it’s a case where there’s no good global default; each implementation is likely to use one or the other consistently, but there isn’t as much consistency between implementations. I saw let’s remove the default.

**Comment by kpcwatson** [comment]
@ktatroe @jbradforddillon are authors responsible for merging? Do PRs need multiple approvals? I don't know if I should be merging these or not.

**Pull request merged**